### ISSUE TITLE
I want to use Record type for keys of Dictionay type.

### ISSUE THREAD

**Issue by yaju** [VBAndCs/VB-Record-Source-Generator#1]
Hi. Great project!

The following Japanese site uses a record type for Dictionary type.
https://note.com/potato_d/n/n99cf0d9edb23

I'm using a translation site, so my English might be weird.

If I run the same thing in VB.NET, even though the keys are the same, they are added without error.

```vb
Public Class TestClass

	Public Sub New()

	End Sub

	Public Sub New(ByVal first As String, ByVal last As String)
		FirstName = first
		LastName = last
	End Sub

	Private _FirstName As String
	Public Property FirstName() As String
		Get
			Return _FirstName
		End Get
		Set(ByVal value As String)
			_FirstName = value
		End Set
	End Property

	Private _LastName As String
	Public Property LastName() As String
		Get
			Return _LastName
		End Get
		Set(ByVal value As String)
			_LastName = value
		End Set
	End Property

	Public Overrides Function Equals(ByVal obj As Object) As Boolean
		Dim comparedObj As TestClass = TryCast(obj, TestClass)

		If comparedObj Is Nothing Then
			Return False
		End If

		Return (comparedObj.FirstName = FirstName) And (comparedObj.LastName = LastName)

	End Function

	Public Overrides Function GetHashCode() As Integer
		Return HashCode.Combine(FirstName, LastName)
	End Function

End Class

Public Class TestRecord
	Public Property FirstName As Object
	Public Property LastName As Object

	Public Sub New(Optional firstName As Object = "", Optional lastName As Object = "")
		_FirstName = firstName
		_LastName = lastName
	End Sub

	Public Shared Function From(anotherRecord As TestRecord, Optional FirstName As Object = Nothing, Optional LastName As Object = Nothing) As TestRecord
		Return New TestRecord(If(FirstName Is Nothing, anotherRecord.FirstName, FirstName), If(LastName Is Nothing, anotherRecord.LastName, LastName))
	End Function

	Public Shared Widening Operator CType(anotherRecord As TestRecord) As (FirstName As Object, LastName As Object)
		Return (anotherRecord.FirstName, anotherRecord.LastName)
	End Operator

	Public Shared Widening Operator CType(fromTuple As (FirstName As Object, LastName As Object)) As TestRecord
		Return New TestRecord(fromTuple.FirstName, fromTuple.LastName)
	End Operator

	Public Shared Function From(fromTuple As (FirstName As Object, LastName As Object)) As TestRecord
		Return New TestRecord(fromTuple.FirstName, fromTuple.LastName)
	End Function

End Class

Module Program
	Sub Main(args As String())

		Dim obj1 As TestClass = New TestClass("Joe", "Blow")
		Dim obj2 As TestClass = New TestClass("Joe", "Blow")

		Try
			Dim dict As New Dictionary(Of TestClass, String) From
			{
				{obj1, "sample1"},
				{obj2, "sample2"}
			}

			Console.WriteLine(dict.Count)
		Catch ex As Exception
			Console.WriteLine(ex.Message)
		End Try

		Dim obj3 As TestRecord = New TestRecord("Joe", "Blow")
		Dim obj4 As TestRecord = New TestRecord("Joe", "Blow")
		Try
			Dim dict As New Dictionary(Of TestRecord, String) From
			{
				{obj3, "sample1"},
				{obj4, "sample2"}
			}
			Console.WriteLine(dict.Count)
		Catch ex As Exception
			Console.WriteLine(ex.Message)
		End Try
	End Sub
End Module
```

If you add overrides for Equal and GetHashCode to TestRecord, you will get the same error.
I need you to think of a fix.

```vb
Public Class TestRecord
	Public Overrides Function Equals(anotherObject) As Boolean
		Dim anotherRecord As TestRecord = TryCast(anotherObject, TestRecord)

		If anotherRecord Is Nothing Then
			Return False
		End If

		Return (anotherRecord.FirstName = FirstName) And (anotherRecord.LastName = LastName)

	End Function

	Public Overrides Function GetHashCode() As Integer
		Return HashCode.Combine(FirstName, LastName)
	End Function
```

**Comment by VBAndCs** [comment]
@yaju 
Did you try to use RecGen to generate your records? This is what this project about.
So, please try to use RecGen and tell me what happens.
Thanks.

**Comment by yaju** [comment]
The same key does not cause an error in RecGen as well.
Therefore, we checked with the class that generated the record in the first submission.

TestRecord.rec

```vb
Public Record TestRecord (
    FirstName As String = "",
    LastName As String = ""
)
```

```vb
Module Program
    Sub Main(args As String())
        Dim obj3 As TestRecord = New TestRecord("Joe", "Blow")
        Dim obj4 As TestRecord = New TestRecord("Joe", "Blow")
        Try
            Dim dict As New Dictionary(Of TestRecord, String) From
            {
                {obj3, "sample1"},
                {obj4, "sample2"}
            }
            Console.WriteLine(dict.Count)　’ result 2
        Catch ex As Exception
            Console.WriteLine(ex.Message)
        End Try
    End Sub
End Module
```

**Comment by VBAndCs** [comment]
@yaju 
I did this to override the GetHashCode method:
```
Partial Public Class TestRecord
	Public Overrides Function GetHashCode() As Integer
		Return HashCode.Combine(FirstName, LastName)
	End Function
End Class
```

Now when I use the record, it causes an exception:
```
       Dim obj3 As  New TestRecord("Joe", "Blow")
        Dim obj4 As New TestRecord("Joe", "Blow")
```

The default hashing depends on the object reference, which is unique, but your hashing seems producing the same value if the two names are identical. Here you should use this:
```
Dim dict As New Dictionary(Of TestRecord, String)
dict(obj3) = "sample1"
dict(obj4) = "sample2"
 ```           
Here the obj4 will overwrite obj3 because they have the same hash (key). Or:
```
Dim dict As New Dictionary(Of TestRecord, String)
dict(obj3) = "sample1"
If not dict.ContainsKey(obj4) Then dict(obj4) = "sample2"
```

This will keep obj3 and reject obj4.
In both ways, the result will be 1.
If this is not what you want, then you should use another hashing to give unique identity for similar records.

Note that vb community has a chat room in Gitter. please join us:
https://gitter.im/VB-NET/community

**Comment by AnthonyDGreen** [comment]
@yaju @VBAndCs 

If I understand this correctly, @yaju is using RecGen but RecGen does not generate `GetHashCode`, only `Equals` and even then it will only generate `Equals` if the properties are marked as `Key` (in the same way that VB anonymous types work, I assume).

Two things:
* @yaju, you must mark `FirstName` and `LastName` as `Key` to get `Equals`, per [the source](https://github.com/VBAndCs/VB-Record-Source-Generator/blob/45f368b42bddf980b2245e4f7f8290f682509ac6/RecordGenerator/RecordParser.vb#L904).
* @VBAndCs RecGen **must** generate `GetHashCode` if it's generating `Equals`. The .NET type system doesn't allow us to express this requirement but [it's documented](https://learn.microsoft.com/en-us/dotnet/api/system.object.gethashcode?view=net-7.0#remarks).

More specifically, two objects which a `.Equals` **must** return the **same** hashcode. This is how dictionaries work, first keys are compared for hashcode equality THEN value equality. If two objects return different hashcodes they are assumed to not be equal. [This line of code](https://github.com/dotnet/runtime/blob/b33aede9f365f9231108c2d9f26ad8955c4e36a1/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/Dictionary.cs#L513) in the implementation of `Dictionary(Of TKey, TValue)` shows this to be the case:

``` C#
if (entries[i].hashCode == hashCode && EqualityComparer<TKey>.Default.Equals(entries[i].key, key))
```

In other words, if they key's hashcodes aren't equal (which by default they won't be with a reference type key, unless the keys are both the same object in memory), `.Equals` is never even called.

This is definitely a bug in RecGen.

**Comment by VBAndCs** [comment]
@AnthonyDGreen 
Thanks for clarifying the issue. I will generate the GetHashCode method to take the key fields into consideration. By the way, the `Record` keyword is a shortcut for `Readonly Key` to have the same effect as C#, so all fields will be Keys and readonly. The need for the `key` is fine tune the fields when we declare a class or a structure, so we can define only some key fields (not all).

**Comment by VBAndCs** [comment]
@yaju 
Until I fix this (which will imply changing the tests and releasing a new NuGet), you can work around this bug by providing the necessary GetHashCode in a partial class, like I should above:
``` vb
Partial Public Class TestRecord
	Public Overrides Function GetHashCode() As Integer
		Return HashCode.Combine(FirstName, LastName)
	End Function
End Class
```

If you have more than 8 keys (which is not expected), use:
```vb
Partial Public Class TestRecord
    Public Overrides Function GetHashCode() As Integer
        dim hash as new HashCode()
        hash.Add(k1)
        hash.Add(k2)
        ' .............
        hash.Add(kn)
        return hash.ToHashCode()
   End Function
End Class
```

which will be the generic form for the generated function in the next version.
Thanks.

**Comment by VBAndCs** [comment]
Fixed in RecGen 2.2
```
PM> Install-Package RecGen
```
